File: 2022/day_2/c/rps.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE_LENGTH 5
#define NUMBER_OF_SHAPES 3
#define PLAYER_SHAPE_STRING "p"
#define OUTCOME_STRING "o"

int get_score_from_player_shape(char letter);
int get_score_from_outcome(char letter);
int get_index_from_letter(char letter);

//   A B C
// X 3 0 6
// Y 6 3 0
// Z 0 6 3

// outcome_score_matrix[opponent_shape][player_shape] = outcome_score
int outcome_score_matrix[NUMBER_OF_SHAPES][NUMBER_OF_SHAPES] = {{3, 6, 0}, {0, 3, 6}, {6, 0, 3}};

//   A B C
// X 3 1 2
// Y 1 2 3
// Z 2 3 1

// player_shape_score_matrix[opponent_shape][outcome] = player_shape_score
int player_shape_score_matrix[NUMBER_OF_SHAPES][NUMBER_OF_SHAPES] = {{3, 1, 2}, {1, 2, 3}, {2, 3, 1}};

int main(int argc, char **argv)
{
    if (argc != 3)
    {
        printf("Usage: ./rps filename second_column_meaning\n");
        return 1;
    }

    char *filename = argv[1];
    FILE *fp = fopen(filename, "r");

    if (fp == NULL)
    {
        printf("Error: could not open file %s\n", filename);
        return 1;
    }

    char *second_column_meaning = argv[2];

    if (strcmp(second_column_meaning, PLAYER_SHAPE_STRING) != 0 && strcmp(second_column_meaning, OUTCOME_STRING) != 0)
    {
        printf("Invalid second_column_meaning passed.  Valid options: '%s' or '%s'\n", PLAYER_SHAPE_STRING, OUTCOME_STRING);
        return 1;
    }

    int total_score = 0;
    char current_line[MAX_LINE_LENGTH];
    int is_second_column_player_shape = strcmp(second_column_meaning, PLAYER_SHAPE_STRING) == 0;

    while (fgets(current_line, MAX_LINE_LENGTH, fp))
    {
        if (is_second_column_player_shape)
        {
            total_score += get_score_from_player_shape(current_line[2]);
            total_score += outcome_score_matrix[get_index_from_letter(current_line[0])][get_index_from_letter(current_line[2])];
        }
        else
        {
            total_score += get_score_from_outcome(current_line[2]);
            total_score += player_shape_score_matrix[get_index_from_letter(current_line[0])][get_index_from_letter(current_line[2])];
        }
    }

    fclose(fp);

    printf("%i", total_score);
}

int get_score_from_player_shape(char letter)
{
    if (letter == 'X')
    {
        return 1;
    }
    else if (letter == 'Y')
    {
        return 2;
    }
    else if (letter == 'Z')
    {
        return 3;
    }
    printf("Unexpected player shape received: %c\n", letter);
    return 0;
}

int get_score_from_outcome(char letter)
{
    if (letter == 'X')
    {
        return 0;
    }
    else if (letter == 'Y')
    {
        return 3;
    }
    else if (letter == 'Z')
    {
        return 6;
    }
    printf("Unexpected outcome received: %c\n", letter);
    return 0;
}

int get_index_from_letter(char letter)
{
    switch (letter)
    {
    case 'A':
    case 'X':
        return 0;
    case 'B':
    case 'Y':
        return 1;
    case 'C':
    case 'Z':
        return 2;
    }
    return 0;
}

File: 2022/day_2/python/rps.py
import sys

# outcome_score_matrix[opponent_shape][player_shape] = outcome_score
outcome_score_matrix = [[ 3, 6, 0 ], [ 0, 3, 6 ], [ 6, 0, 3 ]]

# player_shape_score_matrix[opponent_shape][outcome] = player_shape_score
player_shape_score_matrix = [[ 3, 1, 2 ], [ 1, 2, 3 ], [ 2, 3, 1 ]]

def get_index_from_letter(letter):
    if letter == 'A' or letter == 'X':
        return 0
    elif letter == 'B' or letter == 'Y':
        return 1
    elif letter == 'C' or letter == 'Z':
        return 2
    return 0

def get_score_from_player_shape(letter):
    match letter:
        case 'X':
            return 1
        case 'Y':
            return 2
        case 'Z':
            return 3
    return 0

def get_score_from_outcome(letter):
    match letter:
        case 'X':
            return 0
        case 'Y':
            return 3
        case 'Z':
            return 6
    return 0


if not len(sys.argv) == 3:
    print("Usage: python rps.py filename second_column_meaning")
    sys.exit()

filename = sys.argv[1]
second_column_meaning = sys.argv[2]

if not second_column_meaning == 'p' and not second_column_meaning == 'o':
    print("Invalid second_column_meaning passed.  Valid options: 'p' or 'o'")
    print(second_column_meaning)
    sys.exit()

total_score = 0
is_second_column_player_shape = second_column_meaning == 'p'

with open(filename) as file:
    for line in file:
        if is_second_column_player_shape:
            total_score += get_score_from_player_shape(line[2])
            total_score += outcome_score_matrix[get_index_from_letter(line[0])][get_index_from_letter(line[2])]
        else:
            total_score += get_score_from_outcome(line[2])
            total_score += player_shape_score_matrix[get_index_from_letter(line[0])][get_index_from_letter(line[2])]

print(total_score)